### PR TITLE
[flang] add internal_assoc flag to mark variable captured in internal procedure

### PR DESCRIPTION
**Pull request by jeanPerier**
This patch adds a flag to mark hlfir.declare of host variables that are captured in some internal procedure.

It enables implementing a simple fir.call handling in fir::AliasAnalysis::getModRef leveraging Fortran language specifications and without a data flow analysis.

This will allow implementing an optimization for "array = array_function()" where array storage is passed directly into the hidden result argument to "array_function" when it can be proven that arraY_function does not reference "array".

Captured host variables are very tricky because they may be accessed indirectly in any calls if the internal procedure address was captured via some global procedure pointer. Without flagging them, there is no way around doing a complex inter procedural data flow analysis:
- checking that the call is not made to an internal procedure is not enough because of the possibility of indirect calls made to internal procedures inside the callee.
- checking that the current func.func has no internal procedure is not enough because this would be invalid with inlining when an procedure with internal procedures is inlined inside a procedure without internal procedure.

Next PR using this new flag: https://github.com/llvm/llvm-project/pull/117164

### CODE REVIEW

**Review by razvanlupusoru — APPROVED** [review]
Looks great! Thank you!

File: flang/lib/Lower/ConvertVariable.cpp [hunk 1]
```diff
@@ -1670,6 +1670,25 @@ cuf::DataAttributeAttr Fortran::lower::translateSymbolCUFDataAttribute(
...
+isCapturedInInternalProcedure(Fortran::lower::AbstractConverter &converter,
+                              const Fortran::semantics::Symbol &sym) {
+  const Fortran::lower::pft::FunctionLikeUnit *funit =
+      converter.getCurrentFunctionUnit();
+  if (!funit || funit->getHostAssoc().empty())
+    return false;
+  if (funit->getHostAssoc().isAssociated(sym))
+    return true;
+  // Consider that any capture of a variable that is in an equivalence with the
+  // symbol imply that the storage of the symbol may also be accessed inside
```
Comment (tblah) [inline flang/lib/Lower/ConvertVariable.cpp:1683]:
```suggestion
  // symbol implies that the storage of the symbol may also be accessed inside
```
nit

**Review by tblah — APPROVED** [review]
Thanks!

**Review by clementval — APPROVED** [review]
LGTM

File: flang/lib/Lower/ConvertVariable.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Lower/ConvertVariable.cpp:1683]: Thanks!

**Pull request merged**